linux-gpib-user/doc/man/man3/ibconfig.3
'\" t
.\"     Title: ibconfig
.\"    Author: Frank Mori Hess
.\" Generator: DocBook XSL Stylesheets vsnapshot <http://docbook.sf.net/>
.\"      Date: 10/04/2025
.\"    Manual: 	Traditional API Functions 
.\"    Source: linux-gpib 4.3.7
.\"  Language: English
.\"
.TH "IBCONFIG" "3" "10/04/2025" "linux-gpib 4.3.7" "Traditional API Functions"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
ibconfig \- change configuration (board or device)
.SH "SYNOPSIS"
.sp
.ft B
.nf
#include <gpib/ib\&.h>
.fi
.ft
.HP \w'int\ ibconfig('u
.BI "int ibconfig(int\ " "ud" ", int\ " "option" ", int\ " "setting" ");"
.SH "DESCRIPTION"
.PP
Changes various configuration settings associated with the board or device descriptor
\fIud\fR\&. The
\fIoption\fR
argument specifies the particular setting you wish to modify\&. The
\fIsetting\fR
argument specifies the option\*(Aqs new configuration\&. To query the descriptor\*(Aqs configuration, see
ibask()\&.
.PP
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
.B Table\ \&6.\ \&ibconfig options
.TS
allbox expand tab(:);
lB lB lB lB.
T{
option
T}:T{
value (hexadecimal)
T}:T{
effect
T}:T{
used for board/device
T}
.T&
l l l l
l l l l
l l l l
l l l l
l l l l
l l l l
l l l l
l l l l
l l l l
l l l l
l l l l
l l l l
l l l l
l l l l
l l l l
l l l l
l l l l
l l l l
l l l l
l l l l
l l l l
l l l l
l l l l
l l l l
l l l l
l l l l
l l l l
l l l l
l l l l.
T{
IbcPAD
T}:T{
0x1
T}:T{
Sets GPIB primary address\&.  Same as
	ibpad()
T}:T{
board or device
T}
T{
IbcSAD
T}:T{
0x2
T}:T{
Sets GPIB secondary address\&.  Same as
	ibsad()
T}:T{
board or device
T}
T{
IbcTMO
T}:T{
0x3
T}:T{
Sets timeout for io operations\&.  Same as
	ibmto()\&.
T}:T{
board or device
T}
T{
IbcEOT
T}:T{
0x4
T}:T{
If setting is nonzero, EOI is asserted with last byte on writes\&.
	Same as ibeot()\&.
T}:T{
\ \&
T}
T{
IbcPPC
T}:T{
0x5
T}:T{
Sets parallel poll configuration\&.  Same as
	ibppc()\&.
T}:T{
board
T}
T{
IbcREADDR
T}:T{
0x6
T}:T{
Not implemented, included for compatibility only\&. Device ibrd*
	and ibwrt* calls always re\-address\&. To avoid re\-addressing
	use an initial device read or write followed by board read or
	writes to the controlling minor, ensuring no other intervening
	device access occurs\&.
T}:T{
device
T}
T{
IbcAUTOPOLL
T}:T{
0x7
T}:T{
If setting is nonzero then automatic serial polling is enabled\&.
T}:T{
board
T}
T{
IbcCICPROT
T}:T{
0x8
T}:T{
Useless, included for compatibility only\&.
T}:T{
board
T}
T{
IbcSC
T}:T{
0xa
T}:T{
If setting is nonzero, board becomes system controller\&.
	Same as ibrsc()\&.
T}:T{
board
T}
T{
IbcSRE
T}:T{
0xb
T}:T{
If setting is nonzero then board asserts REN when
	  it becomes sytem controller otherwise REN is not asserted\&.
	  It is nonzero by default\&. See
	ibrsc()\&.
T}:T{
board
T}
T{
IbcEOSrd
T}:T{
0xc
T}:T{
If setting is nonzero then reads are terminated on reception
	of the end\-of\-string character\&.
	See ibeos(),
	in particular the REOS bit\&.
T}:T{
board or device
T}
T{
IbcEOSwrt
T}:T{
0xd
T}:T{
If setting is nonzero then EOI is asserted whenever
	the end\-of\-string character is
	sent\&.  See ibeos(),
	in particular the XEOS bit\&.
T}:T{
board or device
T}
T{
IbcEOScmp
T}:T{
0xe
T}:T{
If setting is nonzero then all 8 bits are used to match
	the end\-of\-string
	character\&.  Otherwise only the least significant 7 bits are used\&.
	See ibeos(),
	in particular the BIN bit\&.
T}:T{
board or device
T}
T{
IbcEOSchar
T}:T{
0xf
T}:T{
Sets the end\-of\-string byte\&.
	See ibeos()\&.
T}:T{
board or device
T}
T{
IbcPP2
T}:T{
0x10
T}:T{
If setting is nonzero then the board is put into
	local parallel poll configure mode (IEEE 488\&.1 PP2 subset), and will not change
	its parallel poll configuration in response to receiving
	\*(Aqparallel poll enable\*(Aq command bytes from the controller\-in\-charge\&.
	Instead, the parallel poll configuration is set locally by doing a board\-level
	call of ibppc()\&.
	A zero value puts the board in remote parallel poll configure mode
	(IEEE 488\&.1 PP1 subset)\&.  IEEE 488\&.2 requires devices to support
	the remote PP1 subset and not the local PP2 subset\&.
	Some older hardware does not support local parallel poll
	configure mode\&.
T}:T{
board
T}
T{
IbcTIMING
T}:T{
0x11
T}:T{
Sets the T1 delay\&.  Use setting of 1 for 2 microseconds, 2 for 500
	nanoseconds, or 3 for 350 nanoseconds\&.  These values are declared in
	the header files as the constants T1_DELAY_2000ns, T1_DELAY_500ns,
	and T1_DELAY_350ns\&.  A 2 microsecond T1 delay is safest, but
	will limit maximum transfer speeds to a few hundred kilobytes
	per second\&.
T}:T{
board
T}
T{
IbcReadAdjust
T}:T{
0x13
T}:T{
If setting is nonzero then byte pairs are automatically swapped
	during reads\&.  Presently, this feature is unimplemented\&.
T}:T{
board or device
T}
T{
IbcWriteAdjust
T}:T{
0x14
T}:T{
If setting is nonzero then byte pairs are automatically swapped
	during writes\&.  Presently, this feature is unimplemented\&.
T}:T{
board or device
T}
T{
IbcEventQueue
T}:T{
0x15
T}:T{
If setting is nonzero then the event queue is enabled\&.  The
	event queue is disabled by default\&.
T}:T{
board
T}
T{
IbcSPollBit
T}:T{
0x16
T}:T{
If the setting is nonzero then the use of the SPOLL bit in ibsta is enabled\&.
T}:T{
board
T}
T{
IbcSendLLO
T}:T{
0x17
T}:T{
If the setting is nonzero then devices connected to this board are automatically
	put into local lockout
	mode when brought online with ibfind() or ibdev()\&.
T}:T{
board
T}
T{
IbcSPollTime
T}:T{
0x18
T}:T{
Sets timeout for serial polls\&.  The setting must be
	between 0 and 17, which correspond to the same time periods as in
	ibtmo()\&.
T}:T{
device
T}
T{
IbcPPollTime
T}:T{
0x19
T}:T{
Sets timeout for parallel polls\&.  The setting must be
	between 0 and 17, which correspond to the same time periods as in
	ibtmo()\&.
T}:T{
board
T}
T{
IbcEndBitIsNormal
T}:T{
0x1a
T}:T{
If setting is nonzero then the END bit of ibsta is set on
	reception of the end\-of\-string character or EOI (default)\&.  Otherwise
	END bit is only set on EOI\&.
T}:T{
board or device
T}
T{
IbcUnAddr
T}:T{
0x1b
T}:T{
If setting is nonzero then UNT (untalk) and UNL (unlisten) commands
	are automatically sent after a completed ibrd* or ibwrt* operation using this descriptor\&.
T}:T{
device
T}
T{
IbcHSCableLength
T}:T{
0x1f
T}:T{
Configures the total cable length in meters for your system, by sending the command
	bytes CFE and CFGn\&.  This is required to enable high speed noninterlocked
	handshaking (a\&.k\&.a\&. HS488) and set associated handshake timings\&.
	Valid \fIsetting\fR values
	are 0 through 15\&.  A value of zero disables noninterlocked handshaking,
	otherwise the value is the total number of meters of cable\&.
T}:T{
board
T}
T{
IbcIst
T}:T{
0x20
T}:T{
Sets the individual status bit, a\&.k\&.a\&. \*(Aqist\*(Aq\&.  Same
	as ibist()\&.
T}:T{
board
T}
T{
IbcRsv
T}:T{
0x21
T}:T{
Sets the current status byte this board will use to respond to
	serial polls\&.  Same as ibrsv()\&.
T}:T{
board
T}
T{
IbcBNA
T}:T{
0x200
T}:T{
Changes the GPIB interface board used to access a device\&.  The
	setting specifies the board index of the new access board\&.  This configuration
	option is similar to ibbna()
	except the new board is specified by its board index instead of a name\&.
T}:T{
device
T}
.TE
.sp 1
.SH "RETURN VALUE"
.PP
The value of
ibsta
is returned\&.
.SH "COPYRIGHT"
.br
Copyright \(co 2003-2006, 2008 Frank Mori Hess
.br


ChangeLog
2018-12-25 09:50  dpenkler

	* [r1806] Fix board type

2018-12-25 09:46  dpenkler

	* [r1805] Fix ni_usb_gpib cut & paste error

2018-12-25 09:42  dpenkler

	* [r1804] Simplify usb udev rules as per Frank's suggestion

2018-12-21 10:44  dpenkler

	* [r1803] Changes for multiple board configuration

	  Update INSTALL
	  Add serial to device environment
	  Print serial in gpib_udev_config
	  Add example rules for agilent

2018-12-15 08:55  fmhess

	* [r1802] Add some examples for configuring multiple boards with
	  different minor numbers.

2018-12-15 08:29  fmhess

	* [r1801] Automatically gibe the correct board type to gpib_config
	  in udev rules

2018-12-15 08:01  fmhess

	* [r1800] rename gpib generic rules so it runs at 98
	  (before the driver-specific fules which run at 99)

2018-12-15 07:52  fmhess

	* [r1799] Add fix for mylog from gpib_udev_config to
	  gpib_udev_fxloader.in. Also add some quoting to
	  ensure whitespace is preserved in log messages.

2018-12-14 16:23  dpenkler

	* [r1798] Fix logging in gpib_udev_config for gpib_config options
	  --

2018-12-14 08:30  fmhess

	* [r1797] Removed some bash-isms so scripts work with posix
	  /bin/sh.

2018-12-14 08:11  fmhess

	* [r1796] Revert commit 1793

2018-12-14 08:07  fmhess

	* [r1795] Add --with-udev-libdir option to configure, that I failed
	  to include in earlier commit.

2018-12-13 17:05  dpenkler

	* [r1794] Really remove unused files

2018-12-13 17:02  dpenkler

	* [r1793] Add UDEV_LIBDIR to Makefile.ams\nRemove unused files

2018-12-13 13:44  fmhess

	* [r1792] Added --with-udev-libdir configure option, defaulted
	  to PREFIX/lib/udev.

2018-12-13 12:24  dpenkler

	* [r1791] Rename and move udevadm wrapper to libdir

2018-12-13 12:17  fmhess

	* [r1790] Add missing define of UDEV_LIB_DIR to prevent files from
	  getting installed into root directory.

2018-12-13 12:13  fmhess

	* [r1789] Replaced udev scripts ni_usb_gpib and agilent_82357a with
	  a single generic script.

2018-12-13 11:53  fmhess

	* [r1788] Really delete usermap files this time.

2018-12-13 11:47  fmhess

	* [r1787] Pull out firmware loading into a single udev script
	  gpib_udev_fxloader.

2018-12-13 11:44  dpenkler

	* [r1786] Use #!/bin/sh instead of #!/bin/bash

2018-12-13 11:11  dpenkler

	* [r1785] Add wrapper script for udevadm

2018-12-12 18:04  dpenkler

	* [r1784] Fix udev rules scripts for agilent usb

2018-12-12 15:43  fmhess

	* [r1783] Generate usb script/rules during "make all" rather
	  than possibly waiting until "make install", where the
	  user may be running as root.

2018-12-12 15:27  fmhess

	* [r1782] Fixed assignment that was supposed to be a comparison

2018-12-12 14:03  fmhess

	* [r1781] Removed obsolete hotplug usermaps.

	  Fixed problems with usb udev rules not working if the driver
	  module was not already loaded when the hardware was plugged
	  in.

	  Made usb drivers respond to the sysfs device path of their
	  usb interface rather than the parent usb device. This is helpful
	  since the driver binds to the interfaces rather than the device.

2018-12-12 06:25  fmhess

	* [r1780] Update comments in agilent_82357a udev script to use
	  new --serial and --sysfs-device-path gpib_config options.

2018-12-10 10:49  fmhess

	* [r1779] Added linux/atomic.h compatibility header.

2018-12-10 10:36  fmhess

	* [r1778] Added a couple more devices to ni_usb_gpib.usermap.

2018-12-10 10:28  fmhess

	* [r1777] Add kernel driver support for selecting hardware by sysfs
	  device path
	  or serial number.

2018-12-10 10:27  fmhess

	* [r1776] Add support for selecting hardware by sysfs device path
	  or serial number. Updated gpib_config docs with
	  missing info on some options. Noted in docs that
	  gpib.conf will be in the sysconfdir, which isn't
	  always /etc/ depending on configuration.

2018-12-10 10:11  fmhess

	* [r1775] Add device ids for various ni adapters to
	  ni_usb_gpib.usermap

2018-12-07 09:12  dpenkler

	* [r1774] Remove dependency on autoconf tools
	  Modify INSTALL instructions
	  Update drivers/gpib/Makefile
	  Add Makefile to invoke kernel build scripts
	  Add version.h to drivers/gpib/include
	  Make definitions of smp_mb__xxx_atomic independent of kernel
	  version
	  Add VERBOSE flag for make, default is VERBOSE=0
	  so that build log is more overseeable.
	  Add HAVE_DEV_OF_NODE workaround for CentOS/RHEL7
	  Fix compile warnings in lpvo

2018-12-06 06:10  fmhess

	* [r1773] Reverted r1766 since it was actually comparing the serial
	  number to the device tree path. Linux doesn't seem to
	  have a universal (for all bus types) way of specifying
	  a piece of hardware by path, so we can't use the
	  device tree path to specify a usb device anyways.

2018-12-05 11:19  dpenkler

	* [r1772] Apply gpib_lines patch from Tomislav Ivek

2018-11-29 08:29  dpenkler

	* [r1771] Change barrier() to mb() to be generic

2018-11-28 19:02  dpenkler

	* [r1770] Add kernel version ifdef for smp atomics

2018-11-28 18:02  dpenkler

	* [r1769] Add compat fixes for kernels < 3.15 and 3.16

2018-11-28 14:36  dpenkler

	* [r1768] Add patch from Tomislav Ivek

	  The current linux-gpib Python class can run into a double-close
	  situation. This bug is related to a design point in Gpib.Gpib
	  which
	  cleans up its GPIB handle only on Python's GC cycle. This is not
	  guaranteed to run when the programmer expects it to and in
	  extreme
	  cases might even leak resources. This is resolved by adding a
	  Gpib.close() method for user-controlled cleanup.

2018-11-28 10:26  fmhess

	* [r1767] Applied patch from Derek Kozel <derek.kozel@gmail.com>:

	  During the linux-gpib-kernel make install process the kernel
	  modules_install command is run bit the DESTDIR argument is not
	  passed
	  in meaning that it will always try to install to the system
	  directory
	  rather than a custom directory. Adding
	  INSTALL_MOD_PATH="$(DESTDIR)" to
	  the command correctly passes the information.

2018-11-16 20:30  fmhess


linux-gpib-user/lib/ibconfig.c
/***************************************************************************
                          lib/ibconfig.c
                             -------------------

    copyright            : (C) 2002 by Frank Mori Hess
    email                : fmhess@users.sourceforge.net
 ***************************************************************************/

/***************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 ***************************************************************************/

#include "ib_internal.h"

static int set_spoll_timeout(ibConf_t *conf, int timeout)
{
	if (timeout < TNONE || timeout > T1000s) {
		setIberr(EARG);
		return -1;
	}

	conf->settings.spoll_usec_timeout = timeout_to_usec(timeout);

	return 0;
}

static int set_ppoll_timeout(ibConf_t *conf, int timeout)
{
	if (timeout < TNONE || timeout > T1000s) {
		setIberr(EARG);
		return -1;
	}

	conf->settings.ppoll_usec_timeout = ppoll_timeout_to_usec(timeout);

	return 0;
}

static int set_t1_delay(ibBoard_t *board, int delay)
{
	unsigned int nano_sec;
	int retval;

	switch (delay) {
		case T1_DELAY_2000ns:
			nano_sec = 2000;
			break;
		case T1_DELAY_500ns:
			nano_sec = 500;
			break;
		case T1_DELAY_350ns:
			nano_sec = 350;
			break;
		default:
			fprintf(stderr, "libgpib: invalid T1 delay selection\n");
			setIberr(EARG);
			return -1;
			break;
	}

	retval = ioctl(board->fileno, IB_T1_DELAY, &nano_sec);
	if (retval < 0)	{
		setIberr(EDVR);
		setIbcnt(errno);
		return -1;
	}

	return 0;
}

static int set_local_parallel_poll_mode(ibBoard_t *board, int local)
{
	int retval;
	short local_mode;

	local_mode = local != 0;

	retval = ioctl(board->fileno, IBPP2_SET, &local_mode);
	if (retval < 0)	{
		setIberr(EDVR);
		setIbcnt(errno);
		return -1;
	}

	return 0;
}

/* Send CFE and CFGn to enable noninterlocked handshaking */
static int set_cable_length (ibConf_t *conf, int num_meters)
{
	int retval;
	uint8_t cmd[2];
	size_t length=0;

	if (num_meters < 0 || num_meters > 15) {
		setIberr (EARG);
		return -1;
	}

	cmd [length++] = CFE;
	if (num_meters)
		cmd [length++] = CFGn(num_meters);
	retval = my_ibcmd (conf, conf->settings.usec_timeout, cmd, length);
	if (retval != length)
		return -1;

	return 0;
}

int ibconfig(int ud, int option, int value)
{
	ibConf_t *conf;
	int retval;

	conf = enter_library(ud);
	if (!conf)
		return exit_library(ud, 1);

	switch (option) {
		case IbcPAD:
			retval = internal_ibpad(conf, value);
			if (retval < 0)
				return exit_library(ud, 1);
			return exit_library(ud, 0);
			break;
		case IbcSAD:
			retval = internal_ibsad(conf, value);
			if (retval < 0)
				return exit_library(ud, 1);
			return exit_library(ud, 0);
			break;
		case IbcTMO:
			retval = internal_ibtmo(conf, value);
			if (retval < 0)
				return exit_library(ud, 1);
			return exit_library(ud, 0);
			break;
		case IbcEOT:
			internal_ibeot(conf, value);
			return exit_library(ud, 0);
			break;
		case IbcEOSrd:
			if (value)
				conf->settings.eos_flags |= REOS;
			else
				conf->settings.eos_flags &= ~REOS;
			return exit_library(ud, 0);
			break;
		case IbcEOSwrt:
			if (value)
				conf->settings.eos_flags |= XEOS;
			else
				conf->settings.eos_flags &= ~XEOS;
			return exit_library(ud, 0);
			break;
		case IbcEOScmp:
			if (value)
				conf->settings.eos_flags |= BIN;
			else
				conf->settings.eos_flags &= ~BIN;
			return exit_library(ud, 0);
			break;
		case IbcEOSchar:
			if ((value & 0xff) != value) {
				setIberr(EARG);
				return exit_library(ud, 1);
			}
			conf->settings.eos = value;
			return exit_library(ud, 0);
			break;
		case IbcReadAdjust:
			// XXX
			if (value) {
				fprintf(stderr, "libgpib: byte swapping on reads not implemented\n");
				setIberr(ECAP);
				return exit_library(ud, 1);
			} else {
				return exit_library(ud, 0);
			}
			break;
		case IbcWriteAdjust:
			// XXX
			if (value) {
				fprintf(stderr, "libgpib: byte swapping on writes not implemented\n");
				setIberr(ECAP);
				return exit_library(ud, 1);
			} else {
				return exit_library(ud, 0);
			}
			break;
		case IbcEndBitIsNormal:
			if (value) {
				return exit_library(ud, 0);
			} else {
				fprintf(stderr, "libgpib: no support for END on EOI only yet \n");
				setIberr(ECAP);
				return exit_library(ud, 1);
			}
			break;
		default:
			break;
	}

	if (conf->is_interface)	{
		switch (option) {
			case IbcPPC:
				retval = internal_ibppc(conf, value);
				if (retval < 0)
					return exit_library(ud, 1);
				return exit_library(ud, 0);
				break;
			case IbcAUTOPOLL:
				retval = configure_autospoll(conf, value);
				if (retval < 0)
					return exit_library(ud, 1);
				return exit_library(ud, 0);
				break;
			case IbcCICPROT:
				// XXX
				if (value) {
					fprintf(stderr, "libgpib: pass control protocol not supported\n");
					setIberr(ECAP);
					return exit_library(ud, 1);
				}else
					return exit_library(ud, 0);
				break;
			case IbcIRQ:
				// XXX
				if (value == 0)	{
					fprintf(stderr, "libgpib: disabling interrupts not supported\n");
					setIberr(ECAP);
					return exit_library(ud, 1);
				} else {
					return exit_library(ud, 0);
				}
				break;
			case IbcSC:
				retval = internal_ibrsc(conf, value);
				if (retval < 0)
					return exit_library(ud, 1);
				else
					return exit_library(ud, 0);
				break;
			case IbcSRE:
				if (value)
					interfaceBoard(conf)->set_ren_on_sc = 1;
				else
					interfaceBoard(conf)->set_ren_on_sc = 0;
				return exit_library(ud, 0);
				break;
			case IbcPP2:
				retval = set_local_parallel_poll_mode(interfaceBoard(conf), value);
				if (retval < 0)
					return exit_library(ud, 1);
				return exit_library(ud, 0);
				break;
			case IbcTIMING:
				if (set_t1_delay(interfaceBoard(conf), value) < 0)
					return exit_library(ud, 1);
				return exit_library(ud, 0);
				break;
			case IbcDMA:
				// XXX
				if (value) {
					return exit_library(ud, 0);
				} else {
					fprintf(stderr, "libgpib: disabling DMA not supported\n");
					setIberr(ECAP);
					return exit_library(ud, 1);
				}
				break;
			case IbcEventQueue:
				if (value) {
					interfaceBoard(conf)->use_event_queue = 1;
				} else {
					interfaceBoard(conf)->use_event_queue = 0;
				}
				return exit_library(ud, 0);
				break;
			case IbcSPollBit:
/*				fprintf(stderr, "libgpib: SPOLL bit support not implemented\n");
				setIberr(ECAP);*/
				/* we currently have some support for SPOLL in the nec7210 driver,
				which is always enabled.  It can only detect the occurance of
				a serial poll if the board was requesting service when the
				poll occurred. */
				return exit_library(ud, 0);
				break;
			case IbcSendLLO:
				// XXX
				if (value) {
					fprintf(stderr, "libgpib: sending local lockout on device open not implemented\n");
					setIberr(ECAP);
					return exit_library(ud, 1);
				} else {
					return exit_library(ud, 0);
				}
				break;
			case IbcPPollTime:
				retval = set_ppoll_timeout(conf, value);
				if (retval < 0) {
					setIberr(EARG);
					return exit_library(ud, 1);
				} else {
					return exit_library(ud, 0);
				}
				break;
			case IbcHSCableLength:
				if (set_cable_length (conf, value) < 0) {
					return exit_library(ud, 1);
				} else {
					return exit_library(ud, 0);
				}
				break;
			case IbcIst:
				retval = internal_ibist(conf, value);
				if (retval < 0)
					return exit_library(ud, 1);
				else
					return exit_library(ud, 0);
				break;
			case IbcRsv:
				retval = internal_ibrsv2(conf, value, value & request_service_bit);
				if (retval < 0)
					return exit_library(ud, 1);
				else
					return exit_library(ud, 0);
				break;
			default:
				break;
		}
	} else {
		switch (option)	{
			case IbcREADDR:
				/* We always re-address.  To support only
				 * readdressing when necessary would require
				 * making the driver keep track of current addressing
				 * state.  Maybe someday, but low priority. */
				if (value)
					conf->settings.readdr = 1;
				else
					conf->settings.readdr = 0;
				return exit_library(ud, 0);
				break;
			case IbcSPollTime:
				retval = set_spoll_timeout(conf, value);
				if (retval < 0)	{
					setIberr(EARG);
					return exit_library(ud, 1);
				} else {
					return exit_library(ud, 0);
				}
				break;
			case IbcUnAddr:
				if (value)
					conf->settings.send_unt_unl = 1;
				else
					conf->settings.send_unt_unl = 0;
				return exit_library(ud, 0);
				break;
			case IbcBNA:
				retval = my_ibbna(conf, value);
				if (retval < 0)
					return exit_library(ud, 1);
				else
					return exit_library(ud, 0);
				break;
			default:
				break;
		}
	}

	setIberr(EARG);
	return exit_library(ud, 1);
}
